File: test/Interpreter/Inputs/SwiftErrorCFunctions/SwiftErrorCFunctions.m
#import "SwiftErrorCFunctions.h"

static NSError *makeError(int code) {
  return [NSError errorWithDomain:@"TestDomain" code:code userInfo:nil];
}

_Bool c_error_zero(_Bool shouldFail, NSError **err) {
  if (shouldFail) {
    if (err) *err = makeError(1);
    return false;
  }
  return true;
}

int32_t c_error_nonzero(int32_t code, NSError **err) {
  if (code != 0) {
    if (err) *err = makeError(code);
    return code;
  }
  return 0;
}

void c_error_nonnull(_Bool shouldFail, NSError **err) {
  if (shouldFail) {
    if (err) *err = makeError(3);
  }
}

void * _Nullable c_error_null(_Bool shouldFail, NSError **err) {
  if (shouldFail) {
    if (err) *err = makeError(4);
    return NULL;
  }
  static int sentinel = 42;
  return &sentinel;
}

void * _Nullable c_error_cf_null(_Bool shouldFail,
    CFErrorRef *err CF_RETURNS_RETAINED) {
  if (shouldFail) {
    if (err) *err = (__bridge_retained CFErrorRef)makeError(5);
    return NULL;
  }
  static int sentinel = 99;
  return &sentinel;
}

void c_error_cf_nonnull(_Bool shouldFail,
    CFErrorRef *err CF_RETURNS_NOT_RETAINED) {
  if (shouldFail) {
    if (err) *err = (__bridge CFErrorRef)makeError(6);
  }
}

_Bool c_error_blocks_both_sides(_Bool shouldFail,
    void (^before)(int), NSError **err, void (^after)(int)) {
  before(10);
  after(20);
  if (shouldFail) {
    if (err) *err = makeError(7);
    return false;
  }
  return true;
}

File: test/Interpreter/swift_error_c_functions.swift
// RUN: %empty-directory(%t)
//
// RUN: %target-clang -fobjc-arc %S/Inputs/SwiftErrorCFunctions/SwiftErrorCFunctions.m -c -o %t/SwiftErrorCFunctions.o
// RUN: %target-build-swift -I %S/Inputs/SwiftErrorCFunctions/ %t/SwiftErrorCFunctions.o %s -o %t/a.out
// RUN: %target-codesign %t/a.out
// RUN: %target-run %t/a.out

// REQUIRES: executable_test
// REQUIRES: objc_interop

import SwiftErrorCFunctions
import Foundation
import StdlibUnittest

var SwiftErrorCFunctionTests = TestSuite("SwiftErrorCFunctions")

// --- zero_result ---

SwiftErrorCFunctionTests.test("zero_result/success") {
  do {
    try c_error_zero(false)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("zero_result/failure") {
  do {
    try c_error_zero(true)
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(1, error.code)
  }
}

// --- nonzero_result ---

SwiftErrorCFunctionTests.test("nonzero_result/success") {
  do {
    try c_error_nonzero(0)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("nonzero_result/failure") {
  do {
    try c_error_nonzero(7)
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(7, error.code)
  }
}

// --- nonnull_error ---

SwiftErrorCFunctionTests.test("nonnull_error/success") {
  do {
    try c_error_nonnull(false)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("nonnull_error/failure") {
  do {
    try c_error_nonnull(true)
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(3, error.code)
  }
}

// --- null_result (NSError) ---

SwiftErrorCFunctionTests.test("null_result/success") {
  do {
    let ptr = try c_error_null(false)
    expectNotNil(ptr)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("null_result/failure") {
  do {
    _ = try c_error_null(true)
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(4, error.code)
  }
}

// --- null_result (CFErrorRef) ---

SwiftErrorCFunctionTests.test("cf_null_result/success") {
  do {
    let ptr = try c_error_cf_null(false)
    expectNotNil(ptr)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("cf_null_result/failure") {
  do {
    _ = try c_error_cf_null(true)
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(5, error.code)
  }
}

// --- nonnull_error (CFErrorRef) ---

SwiftErrorCFunctionTests.test("cf_nonnull_error/success") {
  do {
    try c_error_cf_nonnull(false)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("cf_nonnull_error/failure") {
  do {
    try c_error_cf_nonnull(true)
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(6, error.code)
  }
}

// --- zero_result with blocks on both sides of error param ---

SwiftErrorCFunctionTests.test("blocks_both_sides/success") {
  var beforeArg: Int32 = 0
  var afterArg: Int32 = 0
  do {
    try c_error_blocks_both_sides(false,
                                  { beforeArg = $0 },
                                  { afterArg = $0 })
    expectEqual(10, beforeArg)
    expectEqual(20, afterArg)
  } catch {
    expectUnreachableCatch(error)
  }
}

SwiftErrorCFunctionTests.test("blocks_both_sides/failure") {
  var beforeArg: Int32 = 0
  var afterArg: Int32 = 0
  do {
    try c_error_blocks_both_sides(true,
                                  { beforeArg = $0 },
                                  { afterArg = $0 })
    expectUnreachable()
  } catch let error as NSError {
    expectEqual("TestDomain", error.domain)
    expectEqual(7, error.code)
    // Both blocks ran before the error was raised.
    expectEqual(10, beforeArg)
    expectEqual(20, afterArg)
  }
}

runAllTests()
